runtime/builtin_types.go

```diff
@@ -617,6 +617,85 @@ func builtinSorted(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
 	return result, nil
 }
 
+func builtinSum(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	expectedTypes := []*Type{ObjectType, ObjectType}
+	if argc == 1 {
+		expectedTypes = expectedTypes[:1]
+	}
+	if raised := checkFunctionArgs(f, "sum", args, expectedTypes...); raised != nil {
+		return nil, raised
+	}
+	iter, raised := Iter(f, args[0])
+	if raised != nil {
+		return nil, raised
+	}
+	var result *Object
+	if argc > 1 {
+		if args[1].typ == StrType {
+			return nil, f.RaiseType(TypeErrorType, "sum() can't sum strings [use ''.join(seq) instead]")
+		}
+		result = args[1]
+	} else {
+		result = NewInt(0).ToObject()
+	}
+	if result.typ == IntType {
+		item, raised := Next(f, iter)
+		for ; raised == nil; item, raised = Next(f, iter) {
+			if item.typ == IntType {
+				result = NewInt(toIntUnsafe(result).Value() + toIntUnsafe(item).Value()).ToObject()
+				continue
+			}
+			result, raised = Add(f, result, item)
+			break
+		}
+		if raised != nil {
+			if raised.isInstance(StopIterationType) {
+				f.RestoreExc(nil, nil)
+				return result, nil
+			}
+			return nil, raised
+		}
+	}
+	if result.typ == FloatType {
```

Comment:
My confusion about toIntUnsafe I think demonstrates the down side of this optimization: the code is less understandable. In general I prefer to keep things as simple as possible and optimize only when there's a real performance problem. At that time, I'd want to see benchmarks that demonstrate the poor performance and concrete evidence that the optimization results in a significant improvement, and therefore justifies the additional complexity.

My recommendation is to make this function as simple as possible for now by simply doing Add() in a loop.